ykrt/src/mt.rs

```diff
@@ -379,31 +375,46 @@ impl MTThread {
                 return None;
             }
             PHASE_COMPILING => {
-                let compiling =
-                    unsafe { Box::from_raw((lp & !PHASE_TAG) as *mut CompilingTrace<I>) };
-                match compiling.rcv.try_recv() {
-                    Ok(ct) => {
-                        // FIXME: free up the memory when the trace is no longer used.
-                        let ptr: *mut CompiledTrace<I> = Box::into_raw(Box::new(ct));
-                        let new_state = ptr as usize | PHASE_COMPILED;
-                        loc.state.store(new_state, Ordering::Release);
-
-                        let bct = unsafe { Box::from_raw(ptr) };
-                        let f = unsafe { mem::transmute::<_, fn(&mut I) -> bool>(bct.ptr()) };
-                        mem::forget(bct);
-                        return Some(f);
+                // We need to free the memory allocated earlier. To ensure we don't race with
```

Comment:
I'm probably being daft, but what memory are we freeing again?

Reply:
`let mtx = Arc::new(Mutex::new(None));` (line 359)

Reply:
So it's just the `Arc` wrapper.

Does more than one thread actually need access to what's inside? Could we move the buffer instead of sharing it? Then we wouldn't need another phase?

Reply:
Or is this one of the cases where we can't get Rust to believe us that something lives long enough?

Reply:
`Arc` allocates memory underneath, so it's more than just a wrapper. But "yes".

> Does more than one thread actually need access to what's inside?

Yes, one trace can compile location *L* and another thread might observe that *L* has finished compiling and can be used.

Reply:
Ug. This is hard.

I think I understand. It's a shame we need another phase, but I can't see any other way to do it.